cheap_repr/__init__.py

```diff
@@ -185,7 +185,15 @@ def cheap_repr(x, level=None, target_length=None):
     """
     if level is None:
         level = cheap_repr.max_level
-    x_cls = getattr(x, '__class__', type(x))
+    try:
+        x_cls = getattr(x, '__class__', type(x))
+    except Exception:  # WAT really? yep! i.e. wrapt.ObjectProxy
+        # Otherwise https://github.com/GrahamDumpleton/wrapt/blob/1.13.1/src/wrapt/wrappers.py#L209
+        # causes snooped code to crash during init of this class before it is
+        # in a consistent state for its __getattr__ method to be called. snoop
+        # uses cheap_repr in a tracing hook on every line and does not expect
+        # an exception.
+        return _try_repr(repr, type(x))
```

Comment:
Unfortunately x_cls = type(x) in this situation does not work in the real world scenario I encountered this in.

```
    def test_wrapt_objectproxy(self):
        import wrapt
        class Uhoh(wrapt.ObjectProxy):
            def __init__(self):
                cheap_repr(self)
>       Uhoh()

tests/test_cheap_repr.py:269: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_cheap_repr.py:268: in __init__
    cheap_repr(self)
cheap_repr/__init__.py:204: in cheap_repr
    return _try_repr(func, x, helper)
cheap_repr/__init__.py:217: in _try_repr
    return func(x, *args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

x = <[ValueError('wrapper has not been initialized') raised in repr()] Uhoh object at 0x7f0500c153c0>, helper = <cheap_repr.ReprHelper object at 0x7f0500c15180>

    @register_repr(object)
    @maxparts(60)
    def repr_object(x, helper):
>       s = repr(x)
E       ValueError: wrapper has not been initialized
```

I've updated the test to pull in wrapt to demonstrate that.  My little test apparently wasn't convoluted enough to demonstrate that on its own.

Reply:
Yes, but this is now a different problem. It's always going to be possible for `repr` to raise an exception, but at least now that exception is caught by `_try_repr`. In this case the exception bubbles up from `repr_object`, probably because there's a test that calls `raise_exceptions_from_default_repr()`. In general you should find that the exception is suppressed, and registering a custom repr for `Uhoh` should also work even if it raises an exception.

Reply:
@gpshead ping!

Reply:
sorry, i'm no longer using this code so i'm unlikely to put further work into this PR. i'll close it, someone else can take it over if they see fit.